File: app/src/main/java/com/qbapps/claudeusage/ui/settings/SettingsViewModel.kt
package com.qbapps.claudeusage.ui.settings

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.qbapps.claudeusage.data.local.SecureCredentialStore
import com.qbapps.claudeusage.data.local.UserPreferencesStore
import com.qbapps.claudeusage.domain.model.Organization
import com.qbapps.claudeusage.domain.model.UsageError
import com.qbapps.claudeusage.domain.repository.UsageRepository
import com.qbapps.claudeusage.domain.repository.CodexUsageRepository
import com.qbapps.claudeusage.domain.model.CodexDeviceCode
import com.qbapps.claudeusage.domain.model.GrokDeviceCode
import com.qbapps.claudeusage.domain.repository.GrokUsageRepository
import com.qbapps.claudeusage.data.repository.UsageApiException
import com.qbapps.claudeusage.notification.UsageNotificationHelper
import com.qbapps.claudeusage.worker.WorkManagerScheduler
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.Job
import kotlinx.coroutines.launch
import javax.inject.Inject

data class SettingsUiState(
    val sessionKeyInput: String = "",
    val maskedSessionKey: String? = null,
    val organizations: List<Organization> = emptyList(),
    val selectedOrgId: String? = null,
    val refreshInterval: Int = UserPreferencesStore.DEFAULT_REFRESH_INTERVAL_SECONDS,
    val notifyOnReset: Boolean = true,
    val notifyOnUsageThresholds: Boolean = true,
    val showPersistentNotification: Boolean = false,
    val useRelativeTime: Boolean = true,
    val isValidating: Boolean = false,
    val validationError: String? = null,
    val isKeyValidated: Boolean = false,
    val isCodexConnected: Boolean = false,
    val isCodexConnecting: Boolean = false,
    val codexDeviceCode: CodexDeviceCode? = null,
    val codexSignInError: String? = null,
    val isGrokConnected: Boolean = false,
    val isGrokConnecting: Boolean = false,
    val grokDeviceCode: GrokDeviceCode? = null,
    val grokSignInError: String? = null,
)

@HiltViewModel
class SettingsViewModel @Inject constructor(
    private val credentialStore: SecureCredentialStore,
    private val preferencesStore: UserPreferencesStore,
    private val repository: UsageRepository,
    private val codexRepository: CodexUsageRepository,
    private val grokRepository: GrokUsageRepository,
    private val workManagerScheduler: WorkManagerScheduler,
    private val notificationHelper: UsageNotificationHelper,
) : ViewModel() {

    private val _uiState = MutableStateFlow(SettingsUiState())
    val uiState: StateFlow<SettingsUiState> = _uiState.asStateFlow()
    private var codexLoginJob: Job? = null
    private var grokLoginJob: Job? = null

    init {
        loadExistingSettings()
    }

    fun updateSessionKeyInput(key: String) {
        _uiState.update { it.copy(sessionKeyInput = key, validationError = null) }
    }

    fun connectCodex() {
        codexLoginJob?.cancel()
        codexLoginJob = viewModelScope.launch {
            _uiState.update {
                it.copy(
                    isCodexConnecting = true,
                    codexDeviceCode = null,
                    codexSignInError = null,
                )
            }
            val deviceCode = codexRepository.startDeviceLogin().getOrElse { error ->
                _uiState.update {
                    it.copy(
                        isCodexConnecting = false,
                        codexSignInError = error.message ?: "Codex sign-in could not start.",
                    )
                }
                return@launch
            }
            _uiState.update { it.copy(codexDeviceCode = deviceCode) }

            codexRepository.completeDeviceLogin(deviceCode).fold(
                onSuccess = {
                    val interval = preferencesStore.refreshIntervalSeconds.first()
                    workManagerScheduler.scheduleSync(interval)
                    workManagerScheduler.schedulePeriodicFallback()
                    _uiState.update {
                        it.copy(
                            isCodexConnected = true,
                            isCodexConnecting = false,
                            codexDeviceCode = null,
                            codexSignInError = null,
                        )
                    }
                },
                onFailure = { error ->
                    _uiState.update {
                        it.copy(
                            isCodexConnecting = false,
                            codexDeviceCode = null,
                            codexSignInError = error.message ?: "Codex sign-in failed.",
                        )
                    }
                },
            )
        }
    }

    fun cancelCodexConnection() {
        codexLoginJob?.cancel()
        codexLoginJob = null
        _uiState.update {
            it.copy(
                isCodexConnecting = false,
                codexDeviceCode = null,
                codexSignInError = null,
            )
        }
    }

    fun disconnectCodex() {
        codexLoginJob?.cancel()
        viewModelScope.launch {
            codexRepository.disconnect()
            _uiState.update {
                it.copy(
                    isCodexConnected = false,
                    isCodexConnecting = false,
                    codexDeviceCode = null,
                    codexSignInError = null,
                )
            }
        }
    }

    fun connectGrok() {
        grokLoginJob?.cancel()
        grokLoginJob = viewModelScope.launch {
            _uiState.update {
                it.copy(
                    isGrokConnecting = true,
                    grokDeviceCode = null,
                    grokSignInError = null,
                )
            }
            val deviceCode = grokRepository.startDeviceLogin().getOrElse { error ->
                _uiState.update {
                    it.copy(
                        isGrokConnecting = false,
                        grokSignInError = error.message ?: "Grok sign-in could not start.",
                    )
                }
                return@launch
            }
            _uiState.update { it.copy(grokDeviceCode = deviceCode) }

            grokRepository.completeDeviceLogin(deviceCode).fold(
                onSuccess = {
                    val interval = preferencesStore.refreshIntervalSeconds.first()
                    workManagerScheduler.scheduleSync(interval)
                    workManagerScheduler.schedulePeriodicFallback()
                    _uiState.update {
                        it.copy(
                            isGrokConnected = true,
                            isGrokConnecting = false,
                            grokDeviceCode = null,
                            grokSignInError = null,
                        )
                    }
                },
                onFailure = { error ->
                    _uiState.update {
                        it.copy(
                            isGrokConnecting = false,
                            grokDeviceCode = null,
                            grokSignInError = error.message ?: "Grok sign-in failed.",
                        )
                    }
                },
            )
        }
    }

    fun cancelGrokConnection() {
        grokLoginJob?.cancel()
        grokLoginJob = null
        _uiState.update {
            it.copy(
                isGrokConnecting = false,
                grokDeviceCode = null,
                grokSignInError = null,
            )
        }
    }

    fun disconnectGrok() {
        grokLoginJob?.cancel()
        viewModelScope.launch {
            grokRepository.disconnect()
            _uiState.update {
                it.copy(
                    isGrokConnected = false,
                    isGrokConnecting = false,
                    grokDeviceCode = null,
                    grokSignInError = null,
                )
            }
        }
    }

    fun validateAndSaveKey() {
        val key = _uiState.value.sessionKeyInput.trim()
        if (key.isBlank()) return

        if (!key.startsWith("sk-ant-")) {
            _uiState.update { it.copy(validationError = "Key must start with sk-ant-") }
            return
        }
        if (key.length < 40) {
            _uiState.update { it.copy(validationError = "Key appears too short") }
            return
        }

        viewModelScope.launch {
            _uiState.update { it.copy(isValidating = true, validationError = null) }

            val result = repository.validateSessionKey(key)
            result.fold(
                onSuccess = { orgs ->
                    credentialStore.saveSessionKey(key)
                    _uiState.update {
                        it.copy(
                            isValidating = false,
                            maskedSessionKey = maskKey(key),
                            organizations = orgs,
                            isKeyValidated = true,
                            validationError = null,
                            // Auto-select first org if only one
                            selectedOrgId = if (orgs.size == 1) orgs.first().uuid else it.selectedOrgId,
                        )
                    }
                    // Auto-save if single org
                    if (orgs.size == 1) {
                        selectOrganization(orgs.first())
                    }
                },
                onFailure = { throwable ->
                    val message = when (throwable) {
                        is UsageApiException -> when (throwable.error) {
                            is UsageError.Unauthorized -> "Invalid session key."
                            is UsageError.NetworkError -> "Network error. Check your connection."
                            is UsageError.RateLimited -> "Rate limited. Try again later."
                            is UsageError.ServerError -> "Server error. Try again later."
                            else -> "Validation failed."
                        }
                        else -> "Validation failed: ${throwable.message ?: "Unknown error"}"
                    }
                    _uiState.update {
                        it.copy(isValidating = false, validationError = message)
                    }
                },
            )
        }
    }

    fun selectOrganization(org: Organization) {
        credentialStore.saveOrgId(org.uuid)
        viewModelScope.launch {
            preferencesStore.saveSelectedOrgId(org.uuid)
            // Start background sync now that configuration is complete
            val interval = preferencesStore.refreshIntervalSeconds.first()
            workManagerScheduler.scheduleSync(interval)
            workManagerScheduler.schedulePeriodicFallback()
        }
        _uiState.update { it.copy(selectedOrgId = org.uuid) }
    }

    fun updateRefreshInterval(seconds: Int) {
        _uiState.update { it.copy(refreshInterval = seconds) }
        viewModelScope.launch {
            preferencesStore.saveRefreshInterval(seconds)
            // Reschedule background sync with new interval
            workManagerScheduler.scheduleSync(seconds)
        }
    }

    fun toggleNotifyOnReset(enabled: Boolean) {
        _uiState.update { it.copy(notifyOnReset = enabled) }
        viewModelScope.launch {
            preferencesStore.saveNotifyOnSessionReset(enabled)
        }
    }

    fun toggleNotifyOnUsageThresholds(enabled: Boolean) {
        _uiState.update { it.copy(notifyOnUsageThresholds = enabled) }
        viewModelScope.launch {
            preferencesStore.saveNotifyOnUsageThresholds(enabled)
        }
    }

    fun togglePersistentNotification(enabled: Boolean) {
        _uiState.update { it.copy(showPersistentNotification = enabled) }
        viewModelScope.launch {
            preferencesStore.saveShowPersistentNotification(enabled)
            if (!enabled) {
                notificationHelper.cancelPersistentNotification()
            }
        }
    }

    fun toggleUseRelativeTime(enabled: Boolean) {
        _uiState.update { it.copy(useRelativeTime = enabled) }
        viewModelScope.launch {
            preferencesStore.saveUseRelativeTime(enabled)
        }
    }

    fun clearData() {
        credentialStore.clear()
        workManagerScheduler.cancelAll()
        notificationHelper.cancelPersistentNotification()
        viewModelScope.launch {
            preferencesStore.saveSelectedOrgId(null)
            repository.clearCachedData()
            codexRepository.disconnect()
            grokRepository.disconnect()
        }
        _uiState.update {
            SettingsUiState(
                refreshInterval = it.refreshInterval,
                notifyOnReset = it.notifyOnReset,
                notifyOnUsageThresholds = it.notifyOnUsageThresholds,
            )
        }
    }

    fun clearUsageHistory() {
        viewModelScope.launch {
            repository.clearUsageHistory()
        }
    }

    // ---- internal ----

    private fun loadExistingSettings() {
        val existingKey = credentialStore.getSessionKey()
        val existingOrg = credentialStore.getOrgId()

        _uiState.update {
            it.copy(
                maskedSessionKey = existingKey?.let(::maskKey),
                selectedOrgId = existingOrg,
                isKeyValidated = existingKey != null,
                isCodexConnected = codexRepository.isAuthenticated(),
                isGrokConnected = grokRepository.isAuthenticated(),
            )
        }

        viewModelScope.launch {
            val interval = preferencesStore.refreshIntervalSeconds.first()
            val notifyOnReset = preferencesStore.notifyOnSessionReset.first()
            val notifyOnUsageThresholds = preferencesStore.notifyOnUsageThresholds.first()
            val showPersistent = preferencesStore.showPersistentNotification.first()
            val useRelativeTime = preferencesStore.useRelativeTime.first()
            _uiState.update {
                it.copy(
                    refreshInterval = interval,
                    notifyOnReset = notifyOnReset,
                    notifyOnUsageThresholds = notifyOnUsageThresholds,
                    showPersistentNotification = showPersistent,
                    useRelativeTime = useRelativeTime,
                )
            }
        }
    }

    private fun maskKey(key: String): String {
        if (key.length <= 8) return "****"
        return key.take(7) + "*".repeat(key.length - 11) + key.takeLast(4)
    }
}

File: ios/OpenUsage/ViewModels/SettingsViewModel.swift
import Foundation
import Observation
import OpenUsageCore

/// Port of `ui/settings/SettingsViewModel.kt`.
@MainActor
@Observable
final class SettingsViewModel {
    // Claude
    var sessionKeyInput = ""
    private(set) var maskedSessionKey: String?
    private(set) var organizations: [Organization] = []
    private(set) var selectedOrgId: String?
    private(set) var isValidating = false
    private(set) var validationError: String?
    private(set) var isKeyValidated = false

    // Codex
    private(set) var isCodexConnected = false
    private(set) var isCodexConnecting = false
    private(set) var codexDeviceCode: CodexDeviceCode?
    private(set) var codexSignInError: String?

    // Grok
    private(set) var isGrokConnected = false
    private(set) var isGrokConnecting = false
    private(set) var grokDeviceCode: GrokDeviceCode?
    private(set) var grokSignInError: String?

    // Preferences
    var refreshInterval = PreferencesStore.defaultRefreshIntervalSeconds
    var notifyOnReset = true
    var notifyOnUsageThresholds = true
    var useRelativeTime = true

    private(set) var notificationsAuthorized = true

    private let environment: AppEnvironment
    private var codexLoginTask: Task<Void, Never>?
    private var grokLoginTask: Task<Void, Never>?

    init(environment: AppEnvironment) {
        self.environment = environment
        loadExistingSettings()
    }

    // MARK: - Claude session key

    func validateAndSaveKey() async {
        let key = sessionKeyInput.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !key.isEmpty else { return }
        guard key.hasPrefix("sk-ant-") else {
            validationError = "Key must start with sk-ant-"
            return
        }
        guard key.count >= 40 else {
            validationError = "Key appears too short"
            return
        }

        isValidating = true
        validationError = nil

        do {
            let orgs = try await environment.claudeRepository.validateSessionKey(key)
            environment.credentials.saveSessionKey(key)
            isValidating = false
            maskedSessionKey = Formatters.maskKey(key)
            organizations = orgs
            isKeyValidated = true
            validationError = nil
            sessionKeyInput = ""
            if orgs.count == 1, let only = orgs.first {
                selectOrganization(only)
            }
        } catch let error as UsageError {
            isValidating = false
            validationError = error.validationMessage
        } catch {
            isValidating = false
            validationError = "Validation failed: \(error.localizedDescription)"
        }
    }

    func selectOrganization(_ org: Organization) {
        environment.credentials.saveOrgId(org.uuid)
        environment.preferences.selectedOrgId = org.uuid
        selectedOrgId = org.uuid
        BackgroundRefreshController.schedule(preferredInterval: refreshInterval)
    }

    // MARK: - Codex

    func connectCodex() {
        codexLoginTask?.cancel()
        isCodexConnecting = true
        codexDeviceCode = nil
        codexSignInError = nil

        codexLoginTask = Task { [weak self] in
            guard let self else { return }
            do {
                let code = try await self.environment.codexRepository.startDeviceLogin()
                self.codexDeviceCode = code
                _ = try await self.environment.codexRepository.completeDeviceLogin(code)
                self.isCodexConnected = true
                self.isCodexConnecting = false
                self.codexDeviceCode = nil
                self.codexSignInError = nil
                BackgroundRefreshController.schedule(preferredInterval: self.refreshInterval)
            } catch is CancellationError {
                self.isCodexConnecting = false
                self.codexDeviceCode = nil
            } catch {
                self.isCodexConnecting = false
                self.codexDeviceCode = nil
                self.codexSignInError = Self.message(from: error, fallback: "Codex sign-in failed.")
            }
        }
    }

    func cancelCodexConnection() {
        codexLoginTask?.cancel()
        codexLoginTask = nil
        isCodexConnecting = false
        codexDeviceCode = nil
        codexSignInError = nil
    }

    func disconnectCodex() {
        codexLoginTask?.cancel()
        Task { [weak self] in
            guard let self else { return }
            await self.environment.codexRepository.disconnect()
            self.isCodexConnected = false
            self.isCodexConnecting = false
            self.codexDeviceCode = nil
            self.codexSignInError = nil
        }
    }

    // MARK: - Grok

    func connectGrok() {
        grokLoginTask?.cancel()
        isGrokConnecting = true
        grokDeviceCode = nil
        grokSignInError = nil

        grokLoginTask = Task { [weak self] in
            guard let self else { return }
            do {
                let code = try await self.environment.grokRepository.startDeviceLogin()
                self.grokDeviceCode = code
                _ = try await self.environment.grokRepository.completeDeviceLogin(code)
                self.isGrokConnected = true
                self.isGrokConnecting = false
                self.grokDeviceCode = nil
                self.grokSignInError = nil
                BackgroundRefreshController.schedule(preferredInterval: self.refreshInterval)
            } catch is CancellationError {
                self.isGrokConnecting = false
                self.grokDeviceCode = nil
            } catch {
                self.isGrokConnecting = false
                self.grokDeviceCode = nil
                self.grokSignInError = Self.message(from: error, fallback: "Grok sign-in failed.")
            }
        }
    }

    func cancelGrokConnection() {
        grokLoginTask?.cancel()
        grokLoginTask = nil
        isGrokConnecting = false
        grokDeviceCode = nil
        grokSignInError = nil
    }

    func disconnectGrok() {
        grokLoginTask?.cancel()
        Task { [weak self] in
            guard let self else { return }
            await self.environment.grokRepository.disconnect()
            self.isGrokConnected = false
            self.isGrokConnecting = false
            self.grokDeviceCode = nil
            self.grokSignInError = nil
        }
    }

    // MARK: - Preferences

    func updateRefreshInterval(_ seconds: Int) {
        refreshInterval = seconds
        environment.preferences.refreshIntervalSeconds = seconds
        BackgroundRefreshController.schedule(preferredInterval: seconds)
    }

    func toggleNotifyOnReset(_ enabled: Bool) {
        notifyOnReset = enabled
        environment.preferences.notifyOnSessionReset = enabled
        if enabled { requestNotificationAuthorization() }
    }

    func toggleNotifyOnUsageThresholds(_ enabled: Bool) {
        notifyOnUsageThresholds = enabled
        environment.preferences.notifyOnUsageThresholds = enabled
        if enabled { requestNotificationAuthorization() }
    }

    func toggleUseRelativeTime(_ enabled: Bool) {
        useRelativeTime = enabled
        environment.preferences.useRelativeTime = enabled
    }

    func requestNotificationAuthorization() {
        Task { [weak self] in
            guard let self else { return }
            let granted = await self.environment.notifications.requestAuthorizationIfNeeded()
            self.notificationsAuthorized = granted
        }
    }

    // MARK: - Destructive actions

    func clearUsageHistory() {
        Task { [weak self] in
            await self?.environment.claudeRepository.clearUsageHistory()
        }
    }

    func clearAllData() {
        codexLoginTask?.cancel()
        grokLoginTask?.cancel()
        BackgroundRefreshController.cancelAll()
        environment.notifications.removeAllDelivered()

        Task { [weak self] in
            guard let self else { return }
            await self.environment.refreshService.clearAllData()
            self.sessionKeyInput = ""
            self.maskedSessionKey = nil
            self.organizations = []
            self.selectedOrgId = nil
            self.isKeyValidated = false
            self.validationError = nil
            self.isCodexConnected = false
            self.isGrokConnected = false
            self.codexDeviceCode = nil
            self.grokDeviceCode = nil
            self.loadExistingSettings()
        }
    }

    // MARK: - Internals

    func loadExistingSettings() {
        let existingKey = environment.credentials.sessionKey()
        maskedSessionKey = existingKey.map(Formatters.maskKey)
        isKeyValidated = existingKey != nil
        selectedOrgId = environment.credentials.orgId()

        refreshInterval = environment.preferences.refreshIntervalSeconds
        notifyOnReset = environment.preferences.notifyOnSessionReset
        notifyOnUsageThresholds = environment.preferences.notifyOnUsageThresholds
        useRelativeTime = environment.preferences.useRelativeTime

        Task { [weak self] in
            guard let self else { return }
            self.isCodexConnected = await self.environment.codexRepository.isAuthenticated
            self.isGrokConnected = await self.environment.grokRepository.isAuthenticated
            self.notificationsAuthorized = await self.environment.notifications.isAuthorized()
        }
    }

    /// Refreshes the organization list for an already-saved key so the picker is
    /// populated when Settings is revisited.
    func reloadOrganizationsIfNeeded() {
        guard organizations.isEmpty, let key = environment.credentials.sessionKey() else { return }
        Task { [weak self] in
            guard let self else { return }
            if let orgs = try? await self.environment.claudeRepository.fetchOrganizations(sessionKey: key) {
                self.organizations = orgs
            }
        }
    }

    private static func message(from error: Error, fallback: String) -> String {
        if let provider = error as? ProviderError { return provider.message }
        if let usage = error as? UsageError { return usage.uiMessage }
        return fallback
    }
}
